### PR TITLE
[tune] Ray Tune + Serve end-to-end integration example

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This adds a runnable end-to-end example for using Ray Tune and Ray Serve to the documentation.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: doc/source/tune/_tutorials/overview.rst [hunk 1]
```diff
@@ -73,6 +79,7 @@ Take a look at any of the below tutorials to get started with Tune.
    tune-sklearn.rst
    tune-pytorch-cifar.rst
    tune-pytorch-lightning.rst
+   tune-serve-integration-mnist.rst
```
Comment (anabranch) [inline doc/source/tune/_tutorials/overview.rst:82]: I don't see this `.rst` file - what am I missing?

**Review by bllchmbrs — CHANGES_REQUESTED** [review]

File: doc/source/tune/_tutorials/overview.rst (hunk 1 above)
Comment (krfricke) [reply doc/source/tune/_tutorials/overview.rst:82]: That's the [sphinx-gallery](https://sphinx-gallery.github.io/stable/syntax.html) plugi at work. Basically it takes a formatted python file (`_tutorials/tune-serve-integration-mnist.py`) and creates the RST and an ipython notebook from it. Running `sphinx-build` will first generate these files and place them into `tutorials` (without the underscore), where it is then found by `overview.rst`.

Comment (krfricke) [reply doc/source/tune/_tutorials/overview.rst:82]: We already do this [with the sklearn tutorial](https://docs.ray.io/en/master/tune/tutorials/tune-sklearn.html).

File: doc/source/tune/_tutorials/tune-serve-integration-mnist.py [hunk 2]
```diff
@@ -0,0 +1,645 @@
+# flake8: noqa
+"""
+Model selection and serving with Ray Tune and Ray Serve
+=======================================================
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/tune-serve-integration-mnist.py:4]: How about let's start the first sentence with the goal of this document? Or rather, what you're trying to solve?

File: doc/source/tune/_tutorials/overview.rst [hunk 3]
```diff
@@ -48,6 +48,12 @@ Take a look at any of the below tutorials to get started with Tune.
    :figure: /images/pytorch_lightning_small.png
    :description: :doc:`Tuning PyTorch Lightning modules <tune-pytorch-lightning>`
 
+.. customgalleryitem::
+   :tooltip: Model selection and serving with Ray Tune and Ray Serve
+   :figure: /images/serve.png
```
Comment (simon-mo) [inline doc/source/tune/_tutorials/overview.rst:53]: serve/logo.svg

File: doc/source/tune/_tutorials/tune-serve-integration-mnist.py [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+    checkpoint_path = os.path.join(model_dir, "checkpoint")
+    meta_path = os.path.join(model_dir, "meta.json")
+
+    if os.path.exists(checkpoint_path):
+        shutil.rmtree(checkpoint_path)
+
+    shutil.copytree(checkpoint, checkpoint_path)
+
+    with open(meta_path, "wt") as fp:
+        json.dump(dict(config=config, metrics=metrics), fp)
```
Comment (simon-mo) [inline doc/source/tune/_tutorials/tune-serve-integration-mnist.py:438]: can you move these to a helper function if possible? It's a bit distracting for audience trying figure out which part is doing the serving and which part is just writing checkpoints.

**Review by simon-mo — APPROVED** [review]
LGTM on the Serve side.

File: python/ray/tune/examples/mnist_tune_serve.py [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+
+# Define Ray Serve model,
+class MNISTBackend:
+    def __init__(self, checkpoint_dir, config, metrics):
+        self.checkpoint_dir = checkpoint_dir
+        self.config = config
+        self.metrics = metrics
+
+        # use_cuda = config.get("use_gpu") and torch.cuda.is_available()
+        # self.device = torch.device("cuda" if use_cuda else "cpu")
```
Comment (edoakes) [inline python/ray/tune/examples/mnist_tune_serve.py:116]: any reason this is commented out?

File: python/ray/tune/examples/mnist_tune_serve.py [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+    serve.init()
+    backend_name = "mnist:day_{}".format(day)
+
+    serve.create_backend(backend_name, MNISTBackend, checkpoint_path, config,
+                         metrics)
+    if "mnist" not in serve.list_endpoints():
+        serve.create_endpoint(
+            "mnist", backend=backend_name, route="/mnist", methods=["POST"])
+    else:
+        serve.set_traffic("mnist", {backend_name: 1.0})
```
Comment (edoakes) [inline python/ray/tune/examples/mnist_tune_serve.py:177]: Maybe leave a comment here that you could do incremental rollout by splitting between the new and old models

File: python/ray/tune/examples/mnist_tune_serve.py [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+    checkpoint_path = os.path.join(model_dir, "checkpoint")
+    meta_path = os.path.join(model_dir, "meta.json")
+
+    if os.path.exists(checkpoint_path):
+        shutil.rmtree(checkpoint_path)
+
+    shutil.copytree(checkpoint, checkpoint_path)
+
+    with open(meta_path, "wt") as fp:
+        json.dump(dict(config=config, metrics=metrics), fp)
```
Comment (edoakes) [inline python/ray/tune/examples/mnist_tune_serve.py:166]: How could I do this on multiple nodes? Maybe leave a comment saying checkpoints could easily be synced to S3 instead

**Review by edoakes — APPROVED** [review]
Looks good! Could easily be converted to a short blog as well

File: python/ray/tune/examples/mnist_tune_serve.py (hunk 5 above)
Comment (richardliaw) [reply python/ray/tune/examples/mnist_tune_serve.py:116]: to run on CI iiuc

File: python/ray/tune/examples/mnist_tune_serve.py [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+
+
+def serve_new_model(model_dir, checkpoint, config, metrics, day):
+    print("Serving checkpoint: {}".format(checkpoint))
+
+    if not os.path.exists(model_dir):
+        try:
+            os.mkdir(model_dir, 0o755)
+        except OSError:
+            return False
```
Comment (richardliaw) [inline python/ray/tune/examples/mnist_tune_serve.py:155]: why not just os.make_dirs(model_dir, exist=True)

or something like that?

File: python/ray/tune/examples/mnist_tune_serve.py [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+    best_trial = analysis.get_best_trial("mean_accuracy", "max", "last")
+    best_accuracy = best_trial.metric_analysis["mean_accuracy"]["last"]
+    best_trial_config = best_trial.config
+    best_checkpoint = best_trial.checkpoint.value
+
+    return best_accuracy, best_trial_config, best_checkpoint, N
+
+
+if __name__ == "__main__":
+    """
```
Comment (richardliaw) [inline python/ray/tune/examples/mnist_tune_serve.py:380]: let's put this at top of file?

File: python/ray/tune/examples/mnist_tune_serve.py [hunk 10]
```diff
@@ -0,0 +1,498 @@
...
+
+
+def tune_from_existing(start_model,
+                       start_config,
+                       num_samples=10,
+                       num_epochs=10,
+                       gpus_per_trial=0.,
+                       day=0):
+    data_interface = MNISTDataInterface("/tmp/mnist_data", max_days=10)
+    N = data_interface._get_day_slice(day) - data_interface._get_day_slice(
```
Comment (richardliaw) [inline python/ray/tune/examples/mnist_tune_serve.py:331]: maybe something more intuitive than "N"?

File: doc/source/tune/_tutorials/overview.rst (hunk 3 above)
Comment (krfricke) [reply doc/source/tune/_tutorials/overview.rst:53]: I tried that before but SVGs don't seem to work with `customgalleryitem`:
`OSError: cannot identify image file 'ray/doc/source/serve/logo.svg'`

File: python/ray/tune/examples/mnist_tune_serve.py (hunk 5 above)
Comment (krfricke) [reply python/ray/tune/examples/mnist_tune_serve.py:116]: Thanks for catching this. We'll just go with the fully commented `tune-serve-integration-mnist.py` which is the same code. There, this part is not commented out.

File: python/ray/tune/examples/mnist_tune_serve.py (hunk 9 above)
Comment (krfricke) [reply python/ray/tune/examples/mnist_tune_serve.py:380]: I would actually like to keep it at the end. For one, it is quite conventional to have the main entrypoint at the end of a python script. But the main reason I see is that this code looks a bit cluttered and off-putting compared to the rest of the code, and calls functions would not yet be explained. I think it's much nicer to have everything layed otu/explained before and then have the glue code at the end.

Comment (richardliaw) [reply python/ray/tune/examples/mnist_tune_serve.py:380]: Oh I meant the docstring to go to the head of the file

Comment (krfricke) [reply python/ray/tune/examples/mnist_tune_serve.py:380]: Ah, got it. I think I'll put some of the usage examples from it to the top so people now what the goal for the script is

**Review by bllchmbrs — APPROVED** [review]

**Pull request merged**